### DISCUSSION TITLE
Railsの良いテストについて

### DISCUSSION THREAD

**Discussion by IkumaTadokoro** [roppongirb/discussions#34] [Ideas]
テストをちゃんと書きたいけど、いいサンプルがない（特にRSpec）。
みなさんがどういうところを気をつけているか、どういうことを参考にしているか話したい！

**Comment by yusukesugasawa** [comment]
Roppongi.rb#16 用コメント

**Comment by hatappi** [comment]
## みんなテスト書く時に何つかってる？
rspec

**Reply by hatappi** [reply]
## Why rspec?
- DDD的に仕様を書く流れでテストがかける
  - 他の言語でも rspec っぽいのがある => https://onsi.github.io/ginkgo/
- 構造化できる。`describe, context, it` を書いていくことで１つの文章になる 👍 
- 手に馴染んでいる

テストを見て振る舞いが確認できるのは嬉しい!

`describe, context, it` は英語の理想はありつつ日本語で書いている。
まちまちなことも。

```ruby
describe ".validation"
   context 'when user is not logined' do
      it 'redirects to top page'; end
   end
end
```

**Comment by hatappi** [comment]
## Private method のテストかく
https://t-wada.hatenablog.jp/entry/should-we-test-private-methods

## いらないテストデータの作りすぎを防ぐ
FactoryBot 使っていると associated なリソースが複数があると、余計なリソースが作られすぎてパフォーマンスが落ちちゃうので気をつけている。
- trait 使う
- 使う時に明示的にパラメータを渡すようにしている。

**Comment by hatappi** [comment]
## Shared examples 使ってる？
- 使いすぎ注意
- **使うにしてもその中は before とか余計なリソースを作らずに it とか期待するテスト群をいれるだけにする。**
- 場合わけをする時に便利。e.g. input パターンが複数ある時

**Comment by hatappi** [comment]
## テストケースが多い/テストが重い場合どうしたら良い？
- local でやらずに CI で行うようにしてローカルでの負荷をさげる
- テストが多い場合
  - https://github.com/grosser/parallel_tests を使って CI でガッとやる
  - CI 自体が parallel で実行をサポートしている場合もある (CircleCI)
    - pros: 単純な分割ではなく実行時間をもとにわけるのもある。
    - CI に依存しないのもある: https://docs.knapsackpro.com/ruby/knapsack
- 無駄に input が多いやつではなく境界値テストにしぼる
- １つのテストケースが重い場合あったらどうする？
  - FactoryBot.create_list とかでデータを生成するところがネックになっていた
    - データの生成している部分を見直す
  - **本体のアプリケーションコード自体を見直す** e.g. 分割する

**Comment by hatappi** [comment]
テストが書きづらいというのはアプリケーションの実装に何か見直せる機会があるかも？を疑う良い機会になる
